bn_s_mp_to_decimal_fast.c

```diff
@@ -0,0 +1,234 @@
+/* SPDX-License-Identifier: Unlicense */
+
+/* store a bignum as a decimal ASCII string */
+mp_err s_mp_to_decimal_fast_rec(const mp_int *number, mp_int *nL, mp_int *shiftL, mp_int *mL, int precalc_array_index,
+                                int left,
+                                char **result,
+                                size_t maxlen)
+{
+   mp_int q, nLq, r;
+   mp_err err;
+
+   if (precalc_array_index < 0) {
+      int n = mp_get_i32(number), n2 = n, t = 0, c;
+      char *i = *result;
+      char s[4] = "000";
+
+      while (n) {
+         s[2 - t] = mp_s_rmap[n % 10];
+         t++;
+         n /= 10;
+      }
+
+      if (!left && n2 < 100) {
+         t++;
+         if (n2 < 10) {
+            t++;
+         }
+         if (n2 == 0) {
+            t++;
+         }
+      }
+
+      for (c = 0; c < t; c++) {
+         i[c] = s[3 - t + c];
+      }
+
+      *result += t;
+
+      return MP_OKAY;
+   }
+
+   if ((err = mp_init_multi(&q, &nLq, &r, NULL)) != MP_OKAY) {
+      goto LBL_ERR;
+   }
+   if ((err = mp_mul(number, &mL[precalc_array_index], &q)) != MP_OKAY) {
+      goto LBL_ERR;
+   }
+   if ((err = mp_div_2d(&q, mp_get_i32(&shiftL[precalc_array_index]), &q, NULL)) != MP_OKAY) {
+      goto LBL_ERR;
+   }
+
+   if ((err = mp_mul(&nL[precalc_array_index], &q, &nLq)) != MP_OKAY) {
+      goto LBL_ERR;
+   }
+
+   if ((err = mp_sub(number, &nLq, &r)) != MP_OKAY) {
+      goto LBL_ERR;
+   }
+
+   if (mp_isneg(&r)) {
+      if ((err = mp_sub_d(&q, 1, &q)) != MP_OKAY) {
+         goto LBL_ERR;
+      }
+      if ((err = mp_add(&r, &nL[precalc_array_index], &r)) != MP_OKAY) {
+         goto LBL_ERR;
+      }
+   }
+
+   --precalc_array_index;
+   if (left && mp_iszero(&q)) {
+      if ((err = s_mp_to_decimal_fast_rec(&r, nL, shiftL, mL, precalc_array_index, 1, result, maxlen)) != MP_OKAY) {
+         goto LBL_ERR;
+      }
+   } else {
+      if ((err = s_mp_to_decimal_fast_rec(&q, nL, shiftL, mL, precalc_array_index, left, result, maxlen)) != MP_OKAY) {
+         goto LBL_ERR;
+      }
+      if ((err = s_mp_to_decimal_fast_rec(&r, nL, shiftL, mL, precalc_array_index, 0, result, maxlen)) != MP_OKAY) {
+         goto LBL_ERR;
+      }
+   }
+
+   err = MP_OKAY;
+
+LBL_ERR:
+   mp_clear_multi(&q, &nLq, &r, NULL);
+   return err;
+}
+
+mp_err s_mp_to_decimal_fast(const mp_int *a, char *result, size_t maxlen)
+{
+   mp_int number, n, shift, M, M2, M22, M4, M44;
+   mp_int nL[20], shiftL[20], mL[20];
+   mp_err err;
+   char **result_addr = &result;
+   int precalc_array_index = 1, c;
+
+   if ((err = mp_init_multi(&M2, &M22, &M4, &M44, NULL)) != MP_OKAY) {
+      goto LBL_ERR;
+   }
+
+   if ((err = mp_init_copy(&number, a)) != MP_OKAY) {
+      goto LBL_ERR;
+   }
+   if (mp_isneg(&number)) {
+      if ((err = mp_neg(&number, &number)) != MP_OKAY) {
+         goto LBL_ERR;
+      }
+      result[0] = '-';
+      result_addr += 1;
+   }
+   if ((err = mp_init_set(&n, (mp_digit)1000)) != MP_OKAY) {
```

Comment:
As long as we still support MP_8BIT `mp_digit` can be as small as 7 (seven) bits. Please use e.g.: `mp_set_u32` instead for any (unsigned) value larger than 127.

Reply:
Done in 9d1bae2.